src/neml2/dispatchers/SimpleMPIScheduler.cxx

```diff
@@ -0,0 +1,166 @@
+// By: Argonne National Laboratory
+// OPEN SOURCE LICENSE (MIT)
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "neml2/dispatchers/SimpleMPIScheduler.h"
+#include "neml2/misc/assertions.h"
+
+#include "hwloc.h"
+
+#include <string>
+#include <functional>
+
+namespace neml2
+{
+
+register_NEML2_object(SimpleMPIScheduler);
+
+OptionSet
+SimpleMPIScheduler::expected_options()
+{
+  OptionSet options = WorkScheduler::expected_options();
+  options.doc() =
+      "Dispatch work to a single device selected based on processor ID in given batch sizes.";
+
+  options.set<std::vector<Device>>("devices");
+  options.set("devices").doc() = "List of devices to dispatch work to";
+
+  options.set<std::vector<std::size_t>>("batch_sizes");
+  options.set("batch_sizes").doc() = "List of batch sizes for each device";
+
+  options.set<std::vector<std::size_t>>("capacities") = {};
+  options.set("capacities").doc() = "List of capacities for each device, default to batch_sizes";
+
+  return options;
+}
+
+SimpleMPIScheduler::SimpleMPIScheduler(const OptionSet & options)
+  : WorkScheduler(options),
+    _available_devices(options.get<std::vector<Device>>("devices")),
+    _batch_sizes(options.get<std::vector<std::size_t>>("batch_sizes")),
+    _capacities(options.get("capacities").user_specified()
+                    ? options.get<std::vector<std::size_t>>("capacities")
+                    : _batch_sizes),
+    _comm(TIMPI::Communicator(MPI_COMM_WORLD))
+{
+  neml_assert(_available_devices.size() == _batch_sizes.size(),
+              "Number of batch sizes must match the number of devices.");
+  neml_assert(
+      _available_devices.size() == _capacities.size(),
+      "Number of capacities must match the number of devices and the number of batch sizes.");
+}
+
+void
+SimpleMPIScheduler::setup()
+{
+  WorkScheduler::setup();
+
+  // First pass:
+  // - Prohibit any CPU
```

Comment:
Same question, why prohibiting CPU?

Reply:
We don't have to but you're better off not using a scheduler if you're using this one with the cpu.